src/protocols/tacacs/encode.c

```diff
@@ -23,191 +23,697 @@
+			attr_tacacs_packet->name);
+		return -1;
+	}
+
+	fr_proto_da_stack_build(&da_stack, attr_tacacs_packet);
+	FR_PROTO_STACK_PRINT(&da_stack, 0);
 
 	/*
-	 *	Use memory on the stack, until we know how
-	 *	large the packet will be.
+	 *	Call the struct encoder to do the actual work.
 	 */
-	pkt = (fr_tacacs_packet_t *)data;
+	length_hdr = fr_struct_to_network(buffer, buffer_len, &da_stack, 0, &cursor, NULL, NULL);
 
-	pkt->hdr.ver.major = TAC_PLUS_MAJOR_VER;
+	fr_assert(length_hdr == sizeof(fr_tacacs_packet_hdr_t));
 
-	pkt->hdr.flags = (secret)
-		? TAC_PLUS_ENCRYPTED_MULTIPLE_CONNECTIONS_FLAG
-		: TAC_PLUS_UNENCRYPTED_FLAG;
+	if (length_hdr != sizeof(fr_tacacs_packet_hdr_t)) {
+		fr_strerror_printf("Problems to encode %s using fr_struct_to_network()",
+					attr_tacacs_packet->name);
+		return -1;
+	}
 
-	length_body = 0;
-	for (vp = fr_cursor_init(&cursor, &packet->vps);
-	     vp != NULL;
-	     vp = fr_cursor_next(&cursor)) {
-		VP_VERIFY(vp);
-
-		if (!vp->da->flags.internal) continue;
-
-		if (vp->da == attr_tacacs_version_minor) {
-			pkt->hdr.ver.minor = vp->vp_uint8;
-		} else if (vp->da == attr_tacacs_packet_type) {
-			pkt->hdr.type = vp->vp_uint8;
-		} else if (vp->da == attr_tacacs_sequence_number) {
-			pkt->hdr.seq_no = vp->vp_uint8;
-		} else if (vp->da == attr_tacacs_session_id) {
-			pkt->hdr.session_id = htonl(vp->vp_uint32);
-		} else if (vp->da == attr_tacacs_authentication_status) {
-			pkt->authen.reply.status = vp->vp_uint8;
-			status = vp->vp_uint8;
-		} else if (vp->da == attr_tacacs_authentication_flags) {
-			authen_reply_flags |= vp->vp_uint8;
-		} else if (vp->da == attr_tacacs_authorization_status) {
-			pkt->author.res.status = vp->vp_uint8;
-			status = vp->vp_uint8;
-		} else if (vp->da == attr_tacacs_accounting_status) {
-			pkt->acct.res.status = vp->vp_uint8;
-			status = vp->vp_uint8;
-		} else if (vp->da == attr_tacacs_server_message) {
-			length_body += vp->vp_length;
-			field.server_msg = vp;
-		} else if (vp->da == attr_tacacs_data) {
-			length_body += vp->vp_length;
-			field.data = vp;
+	/*
+	 *	Handle directly in the allocated buffer
+	 */
+	packet = (fr_tacacs_packet_t *)buffer;
+
+	/*
+	 *	Encode 8 octets of various fields not members of STRUCT
+	 */
+	switch (packet->hdr.type) {
+	case TAC_PLUS_AUTHEN:
+		/*
+		 * seq_no
+		 *
+		 * This is the sequence number of the current packet for the current session.
+		 * The first packet in a session MUST have the sequence number 1 and each
+		 * subsequent packet will increment the sequence number by one. Thus clients
+		 * only send packets containing odd sequence numbers, and TACACS+ servers only
+		 * send packets containing even sequence numbers.
+		 */
+		if (packet_is_authen_start_request(packet)) { /* Start */
+			/**
+			 * 4.1. The Authentication START Packet Body
+			 *
+			 * 1 2 3 4 5 6 7 8  1 2 3 4 5 6 7 8  1 2 3 4 5 6 7 8  1 2 3 4 5 6 7 8
+			 * +----------------+----------------+----------------+----------------+
```

Comment:
If you're using the struct decoder would it make more sense to use the struct encoder too?